Test/monadic/MonadicBind.dfy

```diff
@@ -0,0 +1,89 @@
+// RUN: %dafny /compile:3 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+datatype Option<A> = None | Some(get: A) {
+  function method Bind<B>(f: A --> Option<B>): Option<B>
+    requires this == None || f.requires(this.get)
+    reads if this == None then {} else f.reads(this.get)
+  {
+    if this == None then None else f(get)
+  }
+}
```

Comment:
You're intending to implement quite a generic feature, so it would be nice to not only test it for the `Option` monad, but also for some other monad, maybe for the `State` monad? Or what other monads are people intending to use in Dafny?

Reply:
+1 - this also needs to test the case where a type defines both Bind and IsFailure/PropagateFailure/Extract, since the syntax is ambiguous there